lib/cmsis_rtos_v1/cmsis_thread.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <init.h>
+#include <device.h>
+#include "cmsis_thread.h"
+
+struct work_item {
+	struct cmsis_thread *thrd;
+	struct k_delayed_work work;
+};
+static struct work_item worker;
+
+static void work_handler(struct k_work *work)
+{
+	struct work_item *ti = CONTAINER_OF(work, struct work_item, work);
+
+	/*Free the mempool here */
+	k_mem_pool_free(&ti->thrd->block);
+}
+
+static int init_cmsis_rtos_v1_post_kernel(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	k_work_init(&worker.work.work, work_handler);
+	return 0;
+}
+
+SYS_INIT(init_cmsis_rtos_v1_post_kernel,
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+static u32_t zephyr_to_cmsis_priority(s32_t z_prio)
+{
+	return(osPriorityRealtime - z_prio);
+}
+
+static s32_t cmsis_to_zephyr_priority(u32_t c_prio)
+{
+	return(osPriorityRealtime - c_prio);
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	osThreadId thread_id;
+	struct cmsis_thread *cm_thread;
+	void * (*fun_ptr)(void *) = arg3;
+
+	fun_ptr(arg1);
+
+	thread_id = osThreadGetId();
+
+	/* The below operation gets a pointer to the parent structure
+	 * of the thread which also has k_mem_block as one of its elements.
+	 * This needs to be used to free the mem pool.
+	 */
+	cm_thread = CONTAINER_OF(thread_id, struct cmsis_thread, thread);
+
+	worker.thrd = cm_thread;
+	k_work_submit(&worker.work.work);
+}
+
+/* Memory pool for cmsis thread space */
+K_MEM_POOL_DEFINE(cmsis_thread_pool, sizeof(struct cmsis_thread),
+		sizeof(struct cmsis_thread), CONFIG_MAX_CMSIS_THREAD_COUNT, 4);
+
+/**
+ * @brief Create a thread.
+ *
+ * Parameters
+ * [in]	thread_def - thread definition referenced with osThread.
+ * [in]	arg - pointer that is passed to the thread function as start argument.
+ *
+ * Returns
+ * thread ID for reference by other functions or NULL in case of error.
+ */
+osThreadId osThreadCreate(const osThreadDef_t *thread_def, void *arg)
+{
+	s32_t prio;
+	struct k_mem_block block;
+	struct cmsis_thread *cm_thread;
+	k_tid_t tid;
+
+	if (_is_in_isr()) {
+		return NULL;
+	}
+
+	if (!thread_def->stacksize || !thread_def->instances) {
+		return NULL;
+	}
+
+	if (*(thread_def->used_inst_ptr) >= thread_def->instances) {
+		return NULL;
+	}
+
+	(*(thread_def->used_inst_ptr))++;
+	if (k_mem_pool_alloc(&cmsis_thread_pool, &block,
+			     sizeof(struct cmsis_thread), 100) != 0) {
+		/* Insufficient resource to create thread */
+		return NULL;
+	}
+
+	prio = cmsis_to_zephyr_priority(thread_def->tpriority);
+
+	memset(block.data, 0, sizeof(struct cmsis_thread));
+	cm_thread = block.data;
+
+	cm_thread->block.data = block.data;
+	cm_thread->block.id = block.id;
+
+	tid = k_thread_create(&cm_thread->thread,
+			thread_def->stack_mem[*(thread_def->used_inst_ptr) - 1],
```

Comment:
should just be thread_def->stack_mem if you defined it with K_THREAD_STACK_MEMBER